tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala

```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
+  typically at most once per transformation, on the initial input.
+   */
+  /**
+   * Creates a `TBuilderInstruction` from a precomputed `TlaEx`.
+   *
+   * Should be used sparingly, and only for expressions that have already passed static analysis.
+   *
+   * Voids correctness guarantees.
+   */
+  def useTrustedEx(ex: TlaEx): TBuilderInstruction = ex.point[TBuilderInternalState]
+
+  /** x' = y */
+  def primeEq(x: TBuilderInstruction, y: TBuilderInstruction): TBuilderInstruction = eql(prime(x), y)
+
+  type TypedParam = (OperParam, TlaType1)
+
+  /** Parameters to TLA operators can be typed as OperT1 exactly at the top-level, and only if arity > 0 */
+  private def isAcceptableParamType(canContainOper: Boolean): TlaType1 => Boolean = {
+    case FunT1(arg, res)         => isAcceptableParamType(false)(arg) && isAcceptableParamType(false)(res)
+    case SetT1(elem)             => isAcceptableParamType(false)(elem)
+    case SeqT1(elem)             => isAcceptableParamType(false)(elem)
+    case TupT1(elems @ _*)       => elems.forall(isAcceptableParamType(false))
+    case SparseTupT1(fieldTypes) => fieldTypes.values.forall(isAcceptableParamType(false))
+    case RecT1(fieldTypes)       => fieldTypes.values.forall(isAcceptableParamType(false))
+    case OperT1(args, res) =>
+      if (canContainOper)
+        args.nonEmpty &&
+        args.forall(isAcceptableParamType(false)) &&
+        isAcceptableParamType(false)(res)
+      else false
+    case RowT1(fieldTypes, _) => fieldTypes.values.forall(isAcceptableParamType(false))
+    case RecRowT1(row)        => isAcceptableParamType(false)(row)
+    case VariantT1(row)       => isAcceptableParamType(false)(row)
+    case _                    => true
+  }
+
+  private def validateParamType(tp: TypedParam): Unit = {
+    val (OperParam(name, arity), tt) = tp
+    if (!isAcceptableParamType(canContainOper = arity > 0)(tt))
+      throw new TBuilderTypeException(
+          s"Parameter $name type $tt and arity $arity are inconsistent. Parameters have operator types iff their arity is positive."
+      )
+  }
+
+  /**
+   * Operator parameter with type information. Checks that parameters have permissible types.
+   */
+  def param(name: String, tt: TlaType1, arity: Int = 0): TypedParam = {
+    val ret = (OperParam(name, arity), tt)
+    validateParamType(ret)
+    ret
+  }
+
+  /** opName(params[1],...,params[n]) == body */
+  def decl(opName: String, body: TBuilderInstruction, params: TypedParam*): TBuilderOperDeclInstruction = {
+    params.foreach(validateParamType)
+    for {
+      bodyEx <- body
+      // Check param types against their use in body, then mark as bound
+      _ <- params.foldLeft(().point[TBuilderInternalState]) { case (cmp, (OperParam(pName, _), tt)) =>
+        for {
+          _ <- cmp
+          pNameEx <- name(pName, tt) // throws scopeException if param name clashes with param use in body
+          _ <- markAsBound(pNameEx)
+        } yield ()
```

Comment:
Do you really need a for-yield loop here? This looks to me like a lot of advanced features for a couple of side effects

Reply:
Yes, with one computation you could get away with `flatMap`, but both `name` and `markAsBound` are computations. Plus, for expands to chained `flatMap` internally anyway.

Reply:
Also, `for-yield` is hardly an "advanced feature".